crates/ty_python_semantic/src/types.rs

```diff
@@ -1483,6 +1543,164 @@ impl<'db> Type<'db> {
+            Type::NominalInstance(instance) => visitor.try_visit(self, || {
+                instance
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::NominalInstance)
+            }),
+            Type::FunctionLiteral(function) => visitor.try_visit(self, || {
+                function
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::FunctionLiteral)
+            }),
+            Type::PropertyInstance(property) => visitor.try_visit(self, || {
+                property
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::PropertyInstance)
+            }),
+            Type::KnownBoundMethod(method_kind) => visitor.try_visit(self, || {
+                method_kind
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::KnownBoundMethod)
+            }),
+            Type::BoundMethod(method) => visitor.try_visit(self, || {
+                method
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::BoundMethod)
+            }),
+            Type::BoundSuper(bound_super) => visitor.try_visit(self, || {
+                bound_super
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::BoundSuper)
+            }),
+            Type::GenericAlias(generic) => visitor.try_visit(self, || {
+                generic
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::GenericAlias)
+            }),
+            Type::SubclassOf(subclass_of) => visitor.try_visit(self, || {
+                subclass_of
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::SubclassOf)
+            }),
+            Type::TypeVar(bound_typevar) => visitor.try_visit(self, || {
+                Some(Type::TypeVar(
+                    bound_typevar.recursive_type_normalized_impl(db, div, visitor),
+                ))
+            }),
+            Type::KnownInstance(known_instance) => visitor.try_visit(self, || {
+                known_instance
+                    .recursive_type_normalized_impl(db, div, nested, visitor)
+                    .map(Type::KnownInstance)
+            }),
+            Type::TypeIs(type_is) => visitor.try_visit(self, || {
+                let ty = if nested {
+                    type_is
+                        .return_type(db)
+                        .recursive_type_normalized_impl(db, div, true, visitor)?
+                } else {
+                    type_is
+                        .return_type(db)
+                        .recursive_type_normalized_impl(db, div, true, visitor)
+                        .unwrap_or(div)
+                };
+                Some(type_is.with_type(db, ty))
```

Comment:
Here again, it's not immediately clear to me why `TypeIs` is considered a "nesting boundary" (by which I mean, sets `nested` to `true` and and propagates `div` instead of `None` if called non-nested), when many other types are not.